src/trace.ml

```diff
@@ -108,16 +104,58 @@ let write_trace_from_events
           event))
     else events
   in
+  let trace =
+    let%map.Option writer = writer in
+    let base_time =
+      Time_ns.add (Boot_time.time_ns_of_boot_in_perf_time ()) earliest_time
+    in
+    Tracing.Trace.Expert.create ~base_time:(Some base_time) writer
+  in
+  let module Null_writer = struct
```

Comment:
Please define this module top-level unless there's a reason for it to close over anything else. Also, annotate a signature to ensure that errors appear on the module when someone changes the sig instead of where it's used.